### PR TITLE
[Misc] Validate grammar and fail early

### LINKED ISSUE #11118: [Bug]: grammar_is_likely_lark doesn't work correctly
### Your current environment

`grammar_is_likely_lark` identifies valid EBNF grammars as lark like.

### Model Input Dumps

_No response_

### 🐛 Describe the bug

The `grammar_is_likely_lark` is likely too loose, so even valid EBNF grammar is identified as "Lark like" grammar. For example:

```
root ::= (expr "=" term)+
expr  ::= term ([-+*/] term)*
term  ::= num | "(" expr ")"
num   ::= [0-9]+
```

This is a valid EBNF grammar, but since it has `|`, it is identified as lark like grammar.

cc @mgoin 

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by comaniac**
If the grammar is invalid, we now let xGrammar throw RuntimeError when compiling it. However, this happens in logits processor, so the exception is raised from model executor. Since we don't expect model executor to throw any exception now, the exception will crash the engine and kill the worker process.

This PR adds a validation to make sure the grammar is valid when constructing the GrammarConfig to solve this issue.

~~Note that there is another issue with the xgrammar backend that isn't addressed by this PR #11118~~

cc @mgoin

### CODE REVIEW

**Comment by kouroshHakha** [comment]
Hey @comaniac,

1/ Let's add unittests for this. Let's make sure it doesn't diverge from this behavior later on.
2/ We should do a similar thing for json schema as well. Use the following `extra_body` and it will still kill the engine:
`extra_body={"guided_json": {"type": "str"}}`

**Comment by comaniac** [comment]
I don't know where to add unit tests for xgrammar backend. It seems not much unit tests have been added for this component. @mgoin do you have any pointer or should we merge this PR first and make up unit tests later?

Per offline discussion with @mgoin, this PR also fixes the Lark-like issue. I also found another issue that crashes the engine with my fix:
1. Send an invalid grammar. It failed with bad request. However, at this moment the tokenizer data is cached.
2. Send a valid grammar. Since the tokenizer data is cached, we have `encoded_vocab = None`, which results in crash in `get_compiler`. This is because here is a hidden assumption that when `encoded_vocab = None`, the compiler must be initialized already, but this is no longer guaranteed.

The fix in this PR is to make sure `encoded_vocab` would never be `None`. This shouldn't hurt performance because the tokenizer data is cached anyways.

File: vllm/model_executor/guided_decoding/xgrammar_utils.py [hunk 1]
```diff
@@ -26,13 +26,9 @@ def grammar_is_likely_lark(grammar_str: str) -> bool:
...
-            return True
-
-        # Look for Lark-specific features
-        if any(pattern in line for pattern in ['?start:', '|', '~']):
-            return True
+        # Look for GBNF rule definition
+        if '::=' in line:
+            return False
 
     return False
```
Comment (mgoin) [inline vllm/model_executor/guided_decoding/xgrammar_utils.py:33]: I think this should be `return True` now

File: vllm/model_executor/guided_decoding/xgrammar_decoding.py [hunk 2]
```diff
@@ -131,22 +131,22 @@ def from_guided_params(cls,
...
-        # Only get tokenizer data if not already cached
-        if tokenizer_hash in TokenizerDataCache._cache:
-            encoded_vocab = None
-            stop_token_ids = None
-            backend_str = None
-        else:
-            tokenizer_data = TokenizerDataCache.get_tokenizer_data(tokenizer)
-            encoded_vocab = tokenizer_data.encoded_vocab
-            stop_token_ids = tokenizer_data.stop_token_ids
-            backend_str = tokenizer_data.backend_str
```
Comment (comaniac) [inline vllm/model_executor/guided_decoding/xgrammar_decoding.py:143]: `encoded_vocab` cannot be None anymore because the compiler may not be initialized even the tokenizer data is cached if the grammar is invalid. This change shouldn't hurt performance because the tokenizer data is cached anyways.

File: vllm/model_executor/guided_decoding/xgrammar_utils.py (hunk 1 above)
Comment (comaniac) [reply vllm/model_executor/guided_decoding/xgrammar_utils.py:33]: It's "lark like" so we return False when this is GBNF

Comment (mgoin) [reply vllm/model_executor/guided_decoding/xgrammar_utils.py:33]: My issue is there is no case where this function returns True now

Comment (comaniac) [reply vllm/model_executor/guided_decoding/xgrammar_utils.py:33]: Just found that my fix was also wrong lol
Fixed again.

Comment (comaniac) [reply vllm/model_executor/guided_decoding/xgrammar_utils.py:33]: Another wrong fix...sorry about that!

File: vllm/model_executor/guided_decoding/xgrammar_utils.py [hunk 3]
```diff
@@ -26,15 +26,11 @@ def grammar_is_likely_lark(grammar_str: str) -> bool:
...
+        # Look for GBNF rule definition
+        if '::=' in line:
+            return False
 
-        # Look for Lark-specific features
-        if any(pattern in line for pattern in ['?start:', '|', '~']):
-            return True
-
-    return False
+    return True
```
Comment (AlbertoCastelo) [inline vllm/model_executor/guided_decoding/xgrammar_utils.py:33]: This would fix my issue!

**Review by mgoin — APPROVED** [review]
Thanks!
